City-of-Helsinki/helsinki-paatokset
# PP-490: Aggregation and queue changes

This PR improves data aggregation, moving decisions and meeting integrations from JSON files into a Drupal queue.

Error and duplication handling is also improved:
* Items that are added to the queue are checked for duplicates, which should reduce changes of the same failed items and large datasets from clogging up the more important queues.
* Cases where data fetching fails are eventually moved to an error queue instead of being retried forever.

**To test**
* Checkout branch, run `make new`
* Login to the container with `make shell` and pull in test data with `drush mim ahjo_decisionmakers:all`
* Read code changes to make sure there's nothing weird there

**Improvements to decisionmakers list**
* Login with `drush uli` and go to https://helsinki-paatokset.docker.so/en/admin/content/decisionmakers
* The "Sector" and "Organization Name" columns should be visible in the table
* There should be a filter for "Organization name"

**Aggregation changes**
* Run the following commands inside the container to add aggregation data into the processing queue:
  * `drush ap:agg meetings --dataset=latest --start=2023-06-16T00:00:00 --queue -v`
  * `drush ap:agg decisions --dataset=latest --queue -v`
  * `drush ahjo-proxy:aggregate meetings --dataset=cancelled --cancelledonly --queue -v`
* Verify that items were added to the queue: `drush ac:la`
* Process the queue with: `drush queue:run ahjo_api_aggregation_queue -v` (this will take a long time)
* Confirm content was created via aggreation:
  * https://helsinki-paatokset.docker.so/en/admin/content/meetings
  * https://helsinki-paatokset.docker.so/en/admin/content/decisions

**Aggregation duplicate handling**
* Add data to the queues again with:
  * `drush cron` (this will add failed items back to the queue) 
  * `drush ap:agg meetings --dataset=latest --start=2023-06-16T00:00:00 --queue -v`
  * `drush ap:agg decisions --dataset=latest --queue -v`
  * You might get some errors already about not being able to add data to the queues. This is fine.
* Run these commands again:
  * `drush ap:agg meetings --dataset=latest --start=2023-06-16T00:00:00 --queue -v`
  * `drush ap:agg decisions --dataset=latest --queue -v`
  * You should get a lot of errors this time, because the system should not allow you to add duplicates to the queue
  * Verify this by running: `drush ac:la`. The unique item amount should match the amount of total items (or should at least be very close)

**Aggregation error handling**
* Edit this file: `/public/modules/custom/paatokset_ahjo_api/src/AhjoQueueWorkerBase.php`
* Replace line 84 with `$status = 999;`
  * This will cause all data migrations to fail 
* Replace lines 147 and 150 with `$max_time = (int) (new \DateTime('NOW - 3 SECONDS'))->format('U');`
  * This will cause all items to be immediately moved to the retry and error queues (so you don't have to wait 3 days) 
* Run `drush cr`
* Check that the retry and error queues are actually empty:
  * `drush ac:lr`
  * `drush ac:le` 
* Run `drush queue:run ahjo_api_aggregation_queue -v`
  * All the items should fail with a warning and a note  about them being moved
  * Confirm that the items are now in the retry queue: `drush ac:lr`
* Run the retry queue with: `drush queue:run ahjo_api_retry_queue -v`
  * Confirm that the items are now in the error queue: `drush ac:le`
* Run the error queue with `drush queue:run ahjo_api_error_queue -v`
  * This should fail with an error instead of a warning
* Return the failed items back into the queue with `drush cron`
* Add data back to the aggregation queue:
  * `drush ap:agg meetings --dataset=latest --start=2023-06-16T00:00:00 --queue -v`
  * `drush ap:agg decisions --dataset=latest --queue -v`
* Run the queues again:
  * `drush queue:run ahjo_api_aggregation_queue -v
  * `drush queue:run ahjo_api_retry_queue -v
* Check the error queue with `drush ac:le`. There should be no duplicates in this list.
